File: Samples/GSMSample_4_0_Mango_VB/GameStateManagementSample/GameStateManagementSample/Screens/MessageBoxScreen.vb
#Region "File Description"
'-----------------------------------------------------------------------------
' MessageBoxScreen.vb
'
' Microsoft XNA Community Game Platform
' Copyright (C) Microsoft Corporation. All rights reserved.
'-----------------------------------------------------------------------------
#End Region

#Region "Using Statements"
Imports Microsoft.VisualBasic
Imports Microsoft.Xna.Framework
Imports Microsoft.Xna.Framework.Content
Imports Microsoft.Xna.Framework.Graphics
Imports Microsoft.Xna.Framework.Input
Imports GameStateManagement
#End Region

''' <summary>
''' A popup message box screen, used to display "are you sure?"
''' confirmation messages.
''' </summary>
Friend Class MessageBoxScreen
    Inherits GameScreen
#Region "Fields"

    Private message As String
    Private gradientTexture As Texture2D

    Private menuSelect As InputAction
    Private menuCancel As InputAction

#End Region

#Region "Events"

    Public Event Accepted As EventHandler(Of PlayerIndexEventArgs)
    Public Event Cancelled As EventHandler(Of PlayerIndexEventArgs)

#End Region

#Region "Initialization"


    ''' <summary>
    ''' Constructor automatically includes the standard "A=ok, B=cancel"
    ''' usage text prompt.
    ''' </summary>
    Public Sub New(ByVal message As String)
        Me.New(message, True)
    End Sub


    ''' <summary>
    ''' Constructor lets the caller specify whether to include the standard
    ''' "A=ok, B=cancel" usage text prompt.
    ''' </summary>
    Public Sub New(ByVal message As String, ByVal includeUsageText As Boolean)
        Const usageText As String = vbLf & "A button, Space, Enter = ok" & vbLf & "B button, Esc = cancel"

        If includeUsageText Then
            Me.message = message & usageText
        Else
            Me.message = message
        End If

        IsPopup = True

        TransitionOnTime = TimeSpan.FromSeconds(0.2)
        TransitionOffTime = TimeSpan.FromSeconds(0.2)

        menuSelect = New InputAction(New Buttons() {Buttons.A, Buttons.Start}, New Keys() {Keys.Space, Keys.Enter}, True)
        menuCancel = New InputAction(New Buttons() {Buttons.B, Buttons.Back}, New Keys() {Keys.Escape, Keys.Back}, True)
    End Sub


    ''' <summary>
    ''' Loads graphics content for this screen. This uses the shared ContentManager
    ''' provided by the Game class, so the content will remain loaded forever.
    ''' Whenever a subsequent MessageBoxScreen tries to load this same content,
    ''' it will just get back another reference to the already loaded data.
    ''' </summary>
    Public Overrides Sub Activate(ByVal instancePreserved As Boolean)
        If Not instancePreserved Then
            Dim content As ContentManager = ScreenManager.Game.Content
            gradientTexture = content.Load(Of Texture2D)("gradient")
        End If
    End Sub


#End Region

#Region "Handle Input"


    ''' <summary>
    ''' Responds to user input, accepting or cancelling the message box.
    ''' </summary>
    Public Overrides Sub HandleInput(ByVal gameTime As GameTime, ByVal input As InputState)
        Dim playerIndex As PlayerIndex

        ' We pass in our ControllingPlayer, which may either be null (to
        ' accept input from any player) or a specific index. If we pass a null
        ' controlling player, the InputState helper returns to us which player
        ' actually provided the input. We pass that through to our Accepted and
        ' Cancelled events, so they can tell which player triggered them.
        If menuSelect.Evaluate(input, ControllingPlayer, playerIndex) Then
            ' Raise the accepted event, then exit the message box.
            RaiseEvent Accepted(Me, New PlayerIndexEventArgs(playerIndex))

            ExitScreen()
        ElseIf menuCancel.Evaluate(input, ControllingPlayer, playerIndex) Then
            ' Raise the cancelled event, then exit the message box.
            RaiseEvent Cancelled(Me, New PlayerIndexEventArgs(playerIndex))

            ExitScreen()
        End If
    End Sub


#End Region

#Region "Draw"


    ''' <summary>
    ''' Draws the message box.
    ''' </summary>
    Public Overrides Sub Draw(ByVal gameTime As GameTime)
        Dim spriteBatch As SpriteBatch = ScreenManager.SpriteBatch
        Dim font As SpriteFont = ScreenManager.Font

        ' Darken down any other screens that were drawn beneath the popup.
        ScreenManager.FadeBackBufferToBlack(CSng(TransitionAlpha * 2) / 3)

        ' Center the message text in the viewport.
        Dim viewport As Viewport = ScreenManager.GraphicsDevice.Viewport
        Dim viewportSize As New Vector2(viewport.Width, viewport.Height)
        Dim textSize As Vector2 = font.MeasureString(message)
        Dim textPosition As Vector2 = (viewportSize - textSize) / 2

        ' The background includes a border somewhat larger than the text itself.
        Const hPad As Integer = 32
        Const vPad As Integer = 16

        Dim backgroundRectangle As New Rectangle(CInt(textPosition.X) - hPad, CInt(textPosition.Y) - vPad, CInt(textSize.X) + hPad * 2, CInt(textSize.Y) + vPad * 2)

        ' Fade the popup alpha during transitions.
        Dim color As Color = color.White * TransitionAlpha

        spriteBatch.Begin()

        ' Draw the background rectangle.
        spriteBatch.Draw(gradientTexture, backgroundRectangle, color)

        ' Draw the message box text.
        spriteBatch.DrawString(font, message, textPosition, color)

        spriteBatch.End()
    End Sub


#End Region
End Class

File: Samples/RolePlayingGame_4_0_Phone/RolePlayingGame/MenuScreens/MessageBoxScreen.cs
#region File Description
//-----------------------------------------------------------------------------
// MessageBoxScreen.cs
//
// Microsoft XNA Community Game Platform
// Copyright (C) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
#endregion

#region Using Statements
using System;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input.Touch;
using RolePlayingGameData;
#endregion

namespace RolePlaying
{
    /// <summary>
    /// A popup message box screen, used to display "are you sure?"
    /// confirmation messages.
    /// </summary>
    /// <remarks>
    /// Similar to a class found in the Game State Management sample on the 
    /// XNA Creators Club Online website (http://creators.xna.com).
    /// </remarks>
    class MessageBoxScreen : GameScreen
    {
        #region Fields


        string message;

        private Texture2D backgroundTexture;
        private Vector2 backgroundPosition;

        private Texture2D loadingBlackTexture;
        private Rectangle loadingBlackTextureDestination;

        private Texture2D backTexture;
        private Vector2 backPosition;

        private Texture2D selectTexture;
        private Vector2 selectPosition;

        private Vector2 confirmPosition, messagePosition;


        #endregion


        #region Events

        public event EventHandler<EventArgs> Accepted;
        public event EventHandler<EventArgs> Cancelled;

        #endregion


        #region Initialization


        /// <summary>
        /// Constructor lets the caller specify the message.
        /// </summary>
        public MessageBoxScreen(string message)
        {
            this.message = message;

            IsPopup = true;

            TransitionOnTime = TimeSpan.FromSeconds(0.2);
            TransitionOffTime = TimeSpan.FromSeconds(0.2);
        }


        /// <summary>
        /// Loads graphics content for this screen. This uses the shared ContentManager
        /// provided by the Game class, so the content will remain loaded forever.
        /// Whenever a subsequent MessageBoxScreen tries to load this same content,
        /// it will just get back another reference to the already loaded data.
        /// </summary>
        public override void LoadContent()
        {
            ContentManager content = ScreenManager.Game.Content;

            backgroundTexture = content.Load<Texture2D>(@"Textures\MainMenu\Confirm");
            backTexture = content.Load<Texture2D>(@"Textures\Buttons\rpgbtn");
            selectTexture = content.Load<Texture2D>(@"Textures\Buttons\rpgbtn");
            loadingBlackTexture =
                content.Load<Texture2D>(@"Textures\GameScreens\FadeScreen");

            Viewport viewport = ScreenManager.GraphicsDevice.Viewport;
            backgroundPosition = new Vector2(
                (viewport.Width - (backgroundTexture.Width * ScaledVector2.DrawFactor))  / 2,
                (viewport.Height - (backgroundTexture.Height * ScaledVector2.DrawFactor))  / 2);
            loadingBlackTextureDestination = new Rectangle(viewport.X, viewport.Y,
                viewport.Width, viewport.Height);

            backPosition = backgroundPosition + new Vector2(20f,
                backgroundTexture.Height * ScaledVector2.DrawFactor - 110 * ScaledVector2.ScaleFactor);
            selectPosition = backgroundPosition + new Vector2(
                backgroundTexture.Width * ScaledVector2.DrawFactor - 200 * ScaledVector2.ScaleFactor,
                backgroundTexture.Height * ScaledVector2.DrawFactor - 110 * ScaledVector2.ScaleFactor);

            confirmPosition.X = backgroundPosition.X + (backgroundTexture.Width * ScaledVector2.DrawFactor -
                Fonts.HeaderFont.MeasureString("Confirmation").X) / 2f;
            confirmPosition.Y = backgroundPosition.Y + 47 * ScaledVector2.ScaleFactor;

            message = Fonts.BreakTextIntoLines(message, 36, 10);
            messagePosition.X = backgroundPosition.X + (int)((backgroundTexture.Width * ScaledVector2.DrawFactor -
                Fonts.GearInfoFont.MeasureString(message).X) / 2);
            messagePosition.Y = (backgroundPosition.Y * 2) - 20 * ScaledVector2.ScaleFactor;
        }


        #endregion


        #region Handle Input


        /// <summary>
        /// Responds to user input, accepting or cancelling the message box.
        /// </summary>
        public override void HandleInput()
        {
            bool confirmationClicked = false;
            bool backClicked = false;

                
             if (InputManager.IsButtonClicked(new Rectangle(
                    (int)selectPosition.X,
                    (int)selectPosition.Y,
                    (int)(backTexture.Width * ScaledVector2.DrawFactor ),
                    (int)(backTexture.Height * ScaledVector2.DrawFactor ))))
            {
                confirmationClicked = true;
            }

            if (InputManager.IsButtonClicked(new Rectangle
                    ((int)backPosition.X,
                    (int)backPosition.Y,
                    (int)(backTexture.Width * ScaledVector2.DrawFactor ), 
                    (int)(backTexture.Height * ScaledVector2.DrawFactor ))))
            {
                backClicked = true;
            }


            if (confirmationClicked)
            { 
                // Raise the accepted event, then exit the message box.
                if (Accepted != null)
                    Accepted(this, EventArgs.Empty);

                ExitScreen();
            }
            else if (InputManager.IsActionTriggered(InputManager.Action.Back) || backClicked)
            {
                // Raise the cancelled event, then exit the message box.
                if (Cancelled != null)
                    Cancelled(this, EventArgs.Empty);

                ExitScreen();
            }
        }


        #endregion


        #region Draw


        /// <summary>
        /// Draws the message box.
        /// </summary>
        public override void Draw(GameTime gameTime)
        {
            SpriteBatch spriteBatch = ScreenManager.SpriteBatch;

            spriteBatch.Begin();
            var position = new Vector2(loadingBlackTextureDestination.X, loadingBlackTextureDestination.Y);
            spriteBatch.Draw(loadingBlackTexture, position, null,Color.White,0f,
                Vector2.Zero, ScaledVector2.DrawFactor, SpriteEffects.None, 0f);
            spriteBatch.Draw(backgroundTexture, backgroundPosition, null, Color.White, 0f,
                Vector2.Zero, ScaledVector2.DrawFactor, SpriteEffects.None, 0f);
            spriteBatch.Draw(backTexture, backPosition,null, Color.White,0f,
                Vector2.Zero, ScaledVector2.DrawFactor, SpriteEffects.None, 0f);

            spriteBatch.Draw(selectTexture, selectPosition,null, Color.White,0f,
                Vector2.Zero, ScaledVector2.DrawFactor, SpriteEffects.None, 0f);

            string noText = "No";
            Vector2 noTextPosition = Fonts.GetCenterPositionInButton(Fonts.ButtonNamesFont, noText,
                new Rectangle((int)backPosition.X, (int)backPosition.Y,
                    (int)(backTexture.Width), (int)(backTexture.Height )));

            spriteBatch.DrawString(Fonts.ButtonNamesFont, noText,noTextPosition,Color.White);


            string yesText = "Yes";
            Vector2 yesTextPosition = Fonts.GetCenterPositionInButton(Fonts.ButtonNamesFont, yesText,
                new Rectangle((int)selectPosition.X, (int)selectPosition.Y,
                    (int)(selectTexture.Width ), (int)(selectTexture.Height )));

            spriteBatch.DrawString(Fonts.ButtonNamesFont, yesText, yesTextPosition, Color.White);
            spriteBatch.DrawString(Fonts.HeaderFont, "Confirmation", confirmPosition,
                Fonts.CountColor);
            spriteBatch.DrawString(Fonts.GearInfoFont, message, messagePosition, 
                Fonts.CountColor);

            spriteBatch.End();
        }


        #endregion
    }
}
